src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,363 @@
+                Task task = (timerTask == null) ? dequeueTask : await Task.WhenAny(dequeueTask, timerTask).ConfigureAwait(false);
+
+                bool saveBatch = false;
+
+                try
+                {
+                    await task.ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // Happens when node is shutting down or Dispose() is called.
+                    // We want to save whatever is in the batch before exiting the loop.
+                    saveBatch = true;
+
+                    this.logger.LogDebug("Node is shutting down. Save batch.");
+                }
+
+                // Save batch if timer ran out or we've dequeued a new block and reached the consensus tip 
+                // or the max batch size is reached or the node is shutting down.  
+                if (dequeueTask.Status == TaskStatus.RanToCompletion)
+                {
+                    BlockPair item = dequeueTask.Result;
+
+                    // Set the dequeue task to null so it can be assigned on the next iteration.
+                    dequeueTask = null;
+
+                    batch.Add(item);
+
+                    this.currentBatchSizeBytes += item.Block.GetSerializedSize();
+
+                    saveBatch |= (item.ChainedHeader == this.chain.Tip) || (this.currentBatchSizeBytes >= BatchThresholdSizeBytes);
+                }
+                else
+                {
+                    // Will be executed in case timer ran out or node is being shut down.
+                    saveBatch = true;
+                }
+
+                if (saveBatch)
+                {
+                    if (batch.Count != 0)
+                    {
+                        await this.SaveBatchAsync(batch).ConfigureAwait(false);
+
+                        batch.Clear();
+                        this.currentBatchSizeBytes = 0;
+                    }
+
+                    timerTask = null;
+                }
+                else
+                {
+                    // Start timer if it is not started already.
+                    timerTask = timerTask ?? Task.Delay(BatchMaxSaveIntervalSeconds * 1000, this.nodeLifetime.ApplicationStopping);
+                }
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>Checks if repository contains reorged blocks and deletes them; saves batch on top. </summary>
+        /// <param name="batch">List of batched blocks. Cannot be empty.</param>
+        private async Task SaveBatchAsync(List<BlockPair> batch)
+        {
+            this.logger.LogTrace("({0}.{1}:{2})", nameof(batch), nameof(batch.Count), batch.Count);
+
+            List<BlockPair> batchCleared = this.GetBatchWithoutReorgedBlocks(batch);
+
+            ChainedHeader expectedStoreTip = batchCleared.First().ChainedHeader.Previous;
+
+            // Check if block repository contains reorged blocks. If it does - delete them.
```

Comment:
this comment doesn't appear to code below unless expectedStoreTip.hashblock != this.StoreTip.HashBlock means that?  If it does, perhaps that means the variable name of expectedStoreTip could be improved

Reply:
>this comment doesn't appear to code below unless expectedStoreTip.hashblock != this.StoreTip.HashBlock means that?

if `expectedStoreTip.HashBlock != this.storeTip.HashBlock` it means that repo's tip is not something we expect so we need to delete some blocks from them - they are reorged. So yeah, that's exactly what it does. 

>perhaps that means the variable name of expectedStoreTip could be improved

Not sure. This variable contains exactly the store tip that we expect. If it's something else- we need to proceed to removing blocks

Reply:
In that case, is the word expected meant to mean the re-orged or the non re-orged?  perhaps that is what I am not following.

Also the comment says "contains" yet the check is on the hash of the tip

Reply:
>Also the comment says "contains" yet the check is on the hash of the tip

Correct. Because the StoreTip is a ChaniedHeader of the last block that was saved to the database

>In that case, is the word expected meant to mean the re-orged or the non re-orged?

Expected tip- correct one. If expected = actual then don't do anything. If not- remove blocks until highest one is the expected.